### PR TITLE
feat: 4337 function

### PR DESCRIPTION
**Pull request by sanyu1225**
- sendUserOperation
- estimateUserOperationGas
- getUserOperationByHash
- getUserOperationReceipt
- supportedEntryPoints
- generateUserOpHash

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,11966 @@
+{
+  "name": "web3-plugin-eip4337",
+  "version": "0.1.0",
+  "lockfileVersion": 2,
+  "requires": true,
+  "packages": {
```
Comment (jdevcs) [inline package-lock.json]: There are both package-lock and yarn-lock files, you might be using npm and yarn both, I'll recommend to use only one package manager ( in this case yarn ).

File: test/index.test.ts [hunk 2]
```diff
@@ -1,37 +1,82 @@
...
-    it("should call TempltyPlugin test method with expected param", () => {
-      web3Context.EIP4337.test("test-param");
-      expect(consoleSpy).toHaveBeenCalledWith("test-param");
+    it("should call rpcMethods.sendUserOperation with expected parameters", async () => {
+      const entryPoint = "0x636c656d656e74";
+      await web3Context.EIP4337.sendUserOperation(userOperation, entryPoint);
+      expect(sendSpy).toHaveBeenCalledWith(userOperation, entryPoint);
     });
+
+    // it("should call rpcMethods.estimateUserOperationGas with expected parameters", async () => {
```
Comment (jdevcs) [inline test/index.test.ts]: you will also need to add some E2E tests.

**Comment by jdevcs** [comment]
@sanyu1225 are you still working in this PR as its in draft or its ready ?

**Comment by sanyu1225** [comment]
> @sanyu1225 are you still working in this PR as its in draft or its ready ?

@jdevcs I will add some E2E tests, thx.

**Comment by sanyu1225** [comment]
@jdevcs I've finished adding the E2E tests and have reviewed everything. The PR is now updated and ready for review. Let me know if there are any changes needed. Thanks!

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "web3-plugin-eip4337",
-  "version": "0.1.0",
+  "version": "1.0.0",
```
Comment (jdevcs) [inline package.json]: I'll suggest to start 0.1.0 as its first release will be dev / alpha release. and after that in some next PR make it 1.0.0 when its prod ready.

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1,3 @@
 node_modules
-lib
\ No newline at end of file
+lib
+.env
```
Comment (jdevcs) [inline .gitignore]: also filter yarn-error.log files

File: package.json [hunk 5]
```diff
@@ -11,30 +11,45 @@
...
     "build": "tsc --project tsconfig.build.json",
-    "test": "jest --config=./test/jest.config.js"
+    "test": "jest --config=./test/unit/jest.config.js",
+    "pre-black-box": "./scripts/black_box_test_helpers.sh startBackgroundAndPublish",
+    "test:black-box": "yarn pre-black-box && ./scripts/black_box_test_helpers.sh runTests",
+    "test:coverage": "jest --config=./test/unit/jest.config.js --coverage=true --coverage-reporters=text",
+    "test:e2e:chrome": "cypress run --headless --browser chrome",
+    "test:e2e:firefox": "cypress run --headless --browser firefox",
+    "test:e2e:electron": "cypress run --headless --browser electron",
+    "test:all": "yarn && yarn build && yarn test && yarn run test:e2e:firefox && yarn run test:black-box"
```
Comment (jdevcs) [inline package.json:21]: I'll suggest to add tests in github CI so these are invoked in each PR, 
you may see example in existing plugins : 
https://github.com/ChainSafe/web3.js-plugin-chainlink/blob/master/.github/workflows/test.yml
https://web3js.org/plugins

File: src/validator/index.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { UserOperation } from "../type";
+
+export type ValidInputTypes = Uint8Array | bigint | string | number | boolean;
+
+export const isHexStrict = (hex: ValidInputTypes): boolean =>
+  typeof hex === "string" && /^((-)?0x[0-9a-f]+|(0x))$/i.test(hex);
+
```
Comment (Muhammad-Altabba) [inline src/validator/index.ts]: You may use `isHexStrict` from `web3-utils`, if you like:

```suggestion
import { isHexStrict } from "web3-utils";

import { UserOperation } from "../type";

export type ValidInputTypes = Uint8Array | bigint | string | number | boolean;

```

**Review by Muhammad-Altabba — COMMENTED** [review]
Great work!
I have a small suggestion. In addition to Junaid comments.

File: src/validator/index.ts (hunk 6 above)
Comment (sanyu1225) [reply src/validator/index.ts]: Thx, I noticed that the `isHexStrict` function from `web3-utils` was marked as deprecated in the documentation. So, I've switched to using the equivalent function from `web3-validator`instead. [d450e29](https://github.com/web3/web3.js-plugin-eip4337/pull/2/commits/d450e2929e7193a52678c1488c2cd4bfee235c64)

**Comment by jdevcs** [comment]
@sanyu1225 some of CI actions are not passing.

**Comment by sanyu1225** [comment]
> @sanyu1225 some of CI actions are not passing.

Hi, @jdevcs 

The CI black box tests are failing because our latest version isn't published on npm yet. I suggest we merge the current changes and release a new version to fix this. Once updated, we can re-run the CI tests.

Let me know if you're okay with this plan!

Thanks!

**Comment by jdevcs** [comment]
> > @sanyu1225 some of CI actions are not passing.
> 
> Hi, @jdevcs
> 
> The CI black box tests are failing because our latest version isn't published on npm yet. I suggest we merge the current changes and release a new version to fix this. Once updated, we can re-run the CI tests.
> 
> Let me know if you're okay with this plan!
> 
> Thanks!

Hi, I'll suggest to create separate github action for build as compared to tests ( including black box ) . If you want you may do it in another PR?

**Review by jdevcs — APPROVED** [review]

**Comment by sanyu1225** [comment]
> > > @sanyu1225 some of CI actions are not passing.
> > 
> > 
> > Hi, @jdevcs
> > The CI black box tests are failing because our latest version isn't published on npm yet. I suggest we merge the current changes and release a new version to fix this. Once updated, we can re-run the CI tests.
> > Let me know if you're okay with this plan!
> > Thanks!
> 
> Hi, I'll suggest to create separate github action for build as compared to tests ( including black box ) . If you want you may do it in another PR?

Sure, thx

**Pull request merged**